### PR TITLE
fix: warn when training batches are fewer than `accumulate_grad_batches`

### LINKED ISSUE #21808: `accumulate_grad_batches` is silently ignored when epoch length < `accumulate_grad_batches`, because Lightning always steps on the last batch
### Bug description

When a training epoch contains fewer batches than `accumulate_grad_batches`, gradient
accumulation is silently disabled: the optimizer steps on **every** last batch instead
of every `accumulate_grad_batches`-th one. `trainer.global_step` then advances faster.

---

This comes from `TrainingEpochLoop._should_accumulate`:

```python
def _accumulated_batches_reached(self) -> bool:
    """Determine if accumulation will be finished by the end of the current batch."""
    return self.batch_progress.current.ready % self.trainer.accumulate_grad_batches == 0

def _num_ready_batches_reached(self) -> bool:
    """Checks if we are in the last batch or if there are more batches to follow."""
    epoch_finished_on_ready = self.batch_progress.current.ready == self.trainer.num_training_batches
    return epoch_finished_on_ready or self.batch_progress.is_last_batch

def _should_accumulate(self) -> bool:
    """Checks if the optimizer step should be performed or gradients should be accumulated for the current step."""
    accumulation_done = self._accumulated_batches_reached()
    # Lightning steps on the final batch
    is_final_batch = self._num_ready_batches_reached()
    # but the strategy might not
    strategy_accumulates_on_final_batch = self.trainer.strategy.handles_gradient_accumulation or not is_final_batch
    return not accumulation_done and strategy_accumulates_on_final_batch
```

`# Lightning steps on the final batch` is a deliberate design choice (from #2853) meant
to stop accumulation from stalling when an epoch ends mid-cycle (e.g. `accumulate=4` but
only 3 batches left). The unintended consequence: if the epoch is shorter than
`accumulate_grad_batches` in total, *every* batch is the "final batch," so this rule
fires on every step and accumulation never actually happens for the whole run, not
just at the boundary.

This edge case doesn't appear to be documented anywhere (checked the `Trainer`
docstring for `accumulate_grad_batches` and `val_check_interval`, and the gradient
accumulation guide).

### What version are you seeing the problem on?

v2.6

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
``python
import torch
import pytorch_lightning as pl
from torch.utils.data import DataLoader, TensorDataset

class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(4, 1)
        self.val_calls = []

    def training_step(self, batch, batch_idx):
        x, y = batch
        return ((self.layer(x) - y) ** 2).mean()

    def validation_step(self, batch, batch_idx):
        self.val_calls.append(self.global_step)
        return torch.tensor(0.0)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

def run(n_train_samples, batch_size, accumulate_grad_batches, val_every_n_steps, max_steps):
    torch.manual_seed(0)
    train_ds = TensorDataset(torch.randn(n_train_samples, 4), torch.randn(n_train_samples, 1))
    val_ds = TensorDataset(torch.randn(20, 4), torch.randn(20, 1))
    train_dl = DataLoader(train_ds, batch_size=batch_size, shuffle=True)
    val_dl = DataLoader(val_ds, batch_size=batch_size)

    model = Model()
    trainer = pl.Trainer(
        max_steps=max_steps,
        accumulate_grad_batches=accumulate_grad_batches,
        val_check_interval=val_every_n_steps * accumulate_grad_batches,  # docs: int = training batches
        check_val_every_n_epoch=None,
        num_sanity_val_steps=0,
        enable_progress_bar=False,
        enable_checkpointing=False,
        enable_model_summary=False,
        logger=False,
    )
    trainer.fit(model, train_dl, val_dl)
    print(f"n_train_samples={n_train_samples:<3} batches/epoch={len(train_dl):<3} "
          f"-> global_step={model.global_step:<4} val fired at steps={sorted(set(model.val_calls))}")

print("Expected in both cases: global_step reaches 40, validation fires at steps [5,10,...,40]\n")

# Epoch (4 batches) is >= accumulate_grad_batches (4): works as documented.
run(n_train_samples=16, batch_size=4, accumulate_grad_batches=4, val_every_n_steps=5, max_steps=40)

# Epoch (1 batch) is < accumulate_grad_batches (4): accumulation is silently
# defeated because every batch is "the last batch of the epoch".
run(n_train_samples=1, batch_size=4, accumulate_grad_batches=4, val_every_n_steps=5, max_steps=40)


Expected in both cases: global_step reaches 40, validation fires at steps [5,10,...,40]

n_train_samples=16  batches/epoch=4   -> global_step=40   val fired at steps=[5, 10, 15, 20, 25, 30, 35, 40]
n_train_samples=1   batches/epoch=1   -> global_step=40   val fired at steps=[20, 40]


With a 16-sample dataset (4 batches/epoch, `accumulate_grad_batches=4`), one epoch =
one accumulation cycle, and everything lines up as documented.

With a 1-sample dataset (1 batch/epoch, same `accumulate_grad_batches=4`),
`global_step` still reaches 40 in the same number of raw batches, but only because
every batch triggered its own optimizer step — i.e. `accumulate_grad_batches=4` had
zero effect. Validation, which was configured to fire every 5 *optimizer* steps, only
fired at step 20 and 40 instead of 5, 10, 15, 20, ....
```

### Error messages and logs

No error is raised or warned. The run completes silently with the wrong effective
batch size and step cadence, which is what makes this easy to miss.

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):  2.6.1
#- PyTorch Version (e.g., 2.5): 2.12.0+cu130
#- Python version (e.g., 3.12): 3.10.9
#- OS (e.g., Linux): Linux-6.8.0-124-generic-x86_64-with-glibc2.39
#- CUDA/cuDNN version: 13.2
#- GPU models and configuration: nVidia RTX 5090
#- How you installed Lightning(`conda`, `pip`, source): uv pip install
```

</details>

### More info

We hit this in practice training on a deliberately tiny (single-object, duplicated to
just 1 sample) dataset with `accumulate_grad_batches=4`: `val_check_interval` computed
as `val_every_n_steps * accumulate_grad_batches` only fired 4x less often than
configured, and it took a while to trace back to this "steps on the final batch" rule
in `_should_accumulate` rather than a bug in our own step-to-batch conversion.

Would it be reasonable to either:
1. document this interaction on `accumulate_grad_batches` / `val_check_interval`, or
2. emit a `PossibleUserWarning` when `num_training_batches < accumulate_grad_batches`,
   since accumulation is silently a no-op in that case?

Happy to send a PR for either if there's agreement on the approach.

cc @ethanwharris

### PR DESCRIPTION
**Pull request by giacomo-ciro**
## What does this PR do?

When a training epoch contains fewer batches than `accumulate_grad_batches`, Lightning still forces an optimizer step on the last batch of the epoch (a deliberate design choice from #2853, meant to flush leftover gradients when the epoch ends mid-accumulation-cycle). If the whole epoch is shorter than `accumulate_grad_batches`, every batch is the "last batch," so this rule fires on every step and gradient accumulation never actually happens as configured, gradients end up accumulated over fewer batches than intended, with no error or log to indicate it.

As agreed with @deependujha in the issue discussion, this PR adds a `PossibleUserWarning` (instead of a docs-only fix) raised in `FitLoop.setup_data()` when `num_training_batches < accumulate_grad_batches`, following the same pattern as the existing `log_every_n_steps` warning in the same method.

Fixes #21808

This PR does not introduce any breaking changes.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? Yes, see #21808.
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? Not needed. Per @deependujha 's preference, we went with a runtime warning instead of a documentation update.
- Did you write any **new necessary tests**? Yes, added `test_warning_with_dataloader_shorter_than_accumulate_grad_batches`.
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request? None.
- Did you **update the CHANGELOG**? Yes.

</details>

### CODE REVIEW

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -14,10 +14,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Added `log_key_prefix` parameter to `LearningRateMonitor` callback for prefixing logged metric names ([#21612](https://github.com/Lightning-AI/pytorch-lightning/issues/21612))
 
+
+- Added a `PossibleUserWarning` when the number of training batches is smaller than `accumulate_grad_batches`, since Lightning always steps on the last batch of the epoch and gradients end up accumulated over fewer batches than configured ([#21811](https://github.com/Lightning-AI/pytorch-lightning/pull/21811))
+
 - Added support for remote storage (fsspec URLs) when saving and loading distributed checkpoints with `FSDPStrategy` ([#21775](https://github.com/Lightning-AI/pytorch-lightning/pull/21775))
 
 - Added support for remote storage (fsspec URLs) when saving and loading distributed checkpoints with `ModelParallelStrategy` ([#21797](https://github.com/Lightning-AI/pytorch-lightning/issues/21797))
 
+
```
Comment (deependujha) [inline src/lightning/pytorch/CHANGELOG.md]:
```suggestion
```

**Review by deependujha — APPROVED** [review]

File: src/lightning/pytorch/CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added `log_key_prefix` parameter to `LearningRateMonitor` callback for prefixing logged metric names ([#21612](https://github.com/Lightning-AI/pytorch-lightning/issues/21612))
 
+
+- Added a `PossibleUserWarning` when the number of training batches is smaller than `accumulate_grad_batches`, since Lightning always steps on the last batch of the epoch and gradients end up accumulated over fewer batches than configured ([#21811](https://github.com/Lightning-AI/pytorch-lightning/pull/21811))
```
Comment (deependujha) [inline src/lightning/pytorch/CHANGELOG.md]:
```suggestion
- Added a `PossibleUserWarning` when the number of training batches is smaller than `accumulate_grad_batches` ([#21811](https://github.com/Lightning-AI/pytorch-lightning/pull/21811))
```

**Comment by giacomo-ciro** [comment]
@deependujha are the failing checks expected or do I need to do something about it?

**Comment by giacomo-ciro** [comment]
@deependujha all checks had passed. I merged master again because it was updated in the meantime.

**Comment by giacomo-ciro** [comment]
@deependujha shall we merge?

**Comment by giacomo-ciro** [comment]
@deependujha I know you have a lot of PRs on your plate! When you have a moment, could you please approve the workflows so we can merge this? The fix is minimal and won't affect any other tests.

**Review by deependujha — APPROVED** [review]
thanks :)

**Comment by giacomo-ciro** [comment]
@ethanwharris @justusschock @tchaton

It seems like you all are very busy with PRs right now. 

Is it a bad idea to go ahead and merge this since it already has approval, so we do not lose the progress?

**Review by justusschock — APPROVED** [review]
